apps/web/pages/api/basenames/frame/tx.ts

```diff
@@ -0,0 +1,120 @@
+import { encodeFunctionData, namehash } from 'viem';
+import { base } from 'viem/chains';
+import L2ResolverAbi from 'apps/web/src/abis/L2Resolver';
+import RegistrarControllerABI from 'apps/web/src/abis/RegistrarControllerABI';
+import { formatBaseEthDomain } from 'apps/web/src/utils/usernames';
+import {
+  USERNAME_L2_RESOLVER_ADDRESSES,
+  USERNAME_REGISTRAR_CONTROLLER_ADDRESSES,
+} from 'apps/web/src/addresses/usernames';
+
+export type TxFrameStateType = {
+  targetName: string;
+  formattedTargetName: string;
+  targetYears: number;
+  registrationPriceInWei: string;
+  registrationPriceInEth: string;
+};
+
+const RESOLVER_ADDRESS = USERNAME_L2_RESOLVER_ADDRESSES[base.id];
+const REGISTRAR_CONTROLLER_ADDRESS = USERNAME_REGISTRAR_CONTROLLER_ADDRESSES[base.id];
+const NEYNAR_API_KEY = process.env.NEXT_PUBLIC_NEYNAR_API_KEY;
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  if (req.method !== 'POST') {
+    return res.status(405).json({ error: `Tx Screen — Method (${req.method}) Not Allowed` });
+  }
+
+  const body = req.body as FrameRequest;
+  let message;
+  let isValid;
+  let name;
+  let years;
+  let priceInWei;
+  let claimingAddress;
+
+  try {
+    const result = await getFrameMessage(body, {
+      neynarApiKey: NEYNAR_API_KEY,
+    });
+    isValid = result.isValid;
+    message = result.message;
+    if (!isValid) {
+      throw new Error('Message is not valid');
+    }
+    if (!message) {
+      throw new Error('No message received');
+    }
+
+    claimingAddress = message.address as `0x${string}`;
+    if (!claimingAddress) {
+      throw new Error('No address received');
+    }
+
+    const messageState = JSON.parse(
+      decodeURIComponent(message.state?.serialized),
+    ) as TxFrameStateType;
+    if (!messageState) {
+      throw new Error('No message state received');
+    }
+    name = messageState.targetName;
+    years = messageState.targetYears;
+    priceInWei = messageState.registrationPriceInWei;
+  } catch (e) {
+    return res.status(500).json({ error: e });
+  }
+
+  const addressData = encodeFunctionData({
+    abi: L2ResolverAbi,
+    functionName: 'setAddr',
+    args: [namehash(formatBaseEthDomain(name, base.id)), claimingAddress],
+  });
+
+  const nameData = encodeFunctionData({
+    abi: L2ResolverAbi,
+    functionName: 'setName',
+    args: [namehash(formatBaseEthDomain(name, base.id)), formatBaseEthDomain(name, base.id)],
+  });
+
+  const registerRequest = {
+    name,
+    owner: claimingAddress,
+    duration: secondsInYears(years),
+    resolver: RESOLVER_ADDRESS,
+    data: [addressData, nameData],
+    reverseRecord: true,
+  };
+
+  const data = encodeFunctionData({
+    abi: RegistrarControllerABI,
+    functionName: 'register',
+    args: [registerRequest],
+  });
+
+  try {
+    const txData: FrameTransactionResponse = {
+      chainId: `eip155:${base.id}`,
+      method: 'eth_sendTransaction',
+      params: {
+        abi: RegistrarControllerABI,
```

Comment:
I realized after our call that this is kind of a big payload, and you only need the `register` definition, not all the rest of the contract interface. I thiiiiink this should work?

```suggestion
        abi: [{
    inputs: [
      {
        components: [
          { internalType: 'string', name: 'name', type: 'string' },
          { internalType: 'address', name: 'owner', type: 'address' },
          { internalType: 'uint256', name: 'duration', type: 'uint256' },
          { internalType: 'address', name: 'resolver', type: 'address' },
          { internalType: 'bytes[]', name: 'data', type: 'bytes[]' },
          { internalType: 'bool', name: 'reverseRecord', type: 'bool' },
        ],
        internalType: 'struct RegistrarController.RegisterRequest',
        name: 'request',
        type: 'tuple',
      },
    ],
    name: 'register',
    outputs: [],
    stateMutability: 'payable',
    type: 'function',
  }],
```

Reply:
updated